sql,sqlbase,client: bootstrap TableDescriptor timestamps from MVCC

ajwerner opened the pull request:
Using the MVCC timestamp of the value for table descriptors has long been
theorized as the right mechanism to eliminate the need for transactions
which update a table descriptor version to observe their commit timestamp
(see #17698 (comment)).

The challenge was presumed to be the need to expose MVCC timestamps in our
client library. It turns out we seem to do that already (how did nobody know
that?!).

This commit experiments with avoiding using the CommitTimestamp by using the
MVCC timestamp on the read path. It could be cleaned up but seems to work
beautifully. There are still a couple of other places in schema changes which
observe their CommitTimestamp but my sense is this approach will work there
too.

In order to make this setting of the timestamp less of a footgun we add
a `(*Descriptor).Table(hlc.Timestamp)` method which forces anybody who extracts
a `TableDescriptor` from a `Descriptor` to pass in a timestamp which may be
used to set `ModificationTime`. A linter in the second commit enforces this
proper usage.

The below SQL would always fail before this change and now passes:

```
CREATE TABLE foo ( k INT PRIMARY KEY );
BEGIN;
DROP TABLE foo;
<wait a while>
COMMIT;
```

Similarly the TestImportData seems to pass under stressrace with a 5s
`kv.closed_timestamp.target_duration`.

Release justification: Fixes release blocker https://github.com/cockroachdb/cockroach/issues/40504#issuecomment-528612842.

Fixes #21800.
References #37083.

Release note (bug fix): Improve table descriptor update protocol to permit long-running schema change transactions to be pushed and succeed rather than always fail.

Commit:
sql,sqlbase,client: bootstrap TableDescriptor timestamps from MVCC

Using the MVCC timestamp of the value for table descriptors has long been
theorized as the right mechanism to eliminate the need for transactions
which update a table descriptor version to observe their commit timestamp
(see https://github.com/cockroachdb/cockroach/issues/17698#issuecomment-452035873).

The challenge was presumed to be the need to expose MVCC timestamps in our
client library. It turns out we seem to do that already (how did nobody know
that?!).

This commit avoids using the CommitTimestamp by using the MVCC timestamp on
the read path. In order to make this setting of the timestamp less of a footgun
we add a `(*Descriptor).Table(hlc.Timestamp)` method which forces anybody who
extracts a `TableDescriptor` from a `Descriptor` to pass in a timestamp which
may be used to set `ModificationTime` and `CreateAsOfTime`. A linter in the
following commit enforces this proper usage.

The below SQL would always fail before this change and now passes:

```
CREATE TABLE foo ( k INT PRIMARY KEY );
BEGIN;
DROP TABLE foo;
<wait a while>
COMMIT;
```

Similarly the TestImportData seems to pass under stressrace with a 5s
`kv.closed_timestamp.target_duration`.

Release justification: fixes a release blocker and known customer issue.

References #37083.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/40581)
<!-- Reviewable:end -->

petermattis reviewed (commented):
> The challenge was presumed to be the need to expose MVCC timestamps in our
> client library. It turns out we seem to do that already (how did nobody know
> that?!).

&lt;Raises hand sheepishly&gt;, though note that `ScanRequest` will only return timestamps if the slower `KEY_VALUES` format is used.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-:-LoGnQWx39jsXbuTjmuJ:b-t49edz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @lucy-zhang)

___
*[pkg/ccl/changefeedccl/table_history.go, line 286 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoGqJYg65Avlfy4sWVn:-LoGqJYg65Avlfy4sWVo:btk64dn) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae3b2a00366d69614cdea0c2bffac5655b1fb4d/pkg/ccl/changefeedccl/table_history.go#L286)):*
> ```Go
> 				}
> 				if tableDesc := desc.GetTable(); tableDesc != nil {
> 					tableDesc.MaybeSetModificationTimeFromMVCCTimestamp(ctx, it.UnsafeKey().Timestamp)
> ```

How bad is it going to be if you miss one of the places that table descriptors are decoded? I'm wondering if there is anything that could be done to make this less error prone.  For example, could we change the signature for `desc.GetTable()` to pass in a timestamp which would force you to update all of the callers. Seems like there are too many call sites for that, but maybe that sparks a better idea.
___
*[pkg/internal/client/db.go, line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoGmq0l9xKHMzAfH2e5:-LoGmq0l9xKHMzAfH2e6:bv79kav) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae3b2a00366d69614cdea0c2bffac5655b1fb4d/pkg/internal/client/db.go#L34)):*
> ```Go
> type KeyValue struct {
> 	Key roachpb.Key
> 	// TODO(ajwerner): figure out why it says timestamp will always be zero
> ```

When using `ScanFormat.BATCH_RESPONSE`, the `ScanResponse.BatchResponses` field will contain encoded key/value pairs which do not have a timestamp. See `enginepb.ScanDecodeKeyValueNoTS`. I think the value in `GetResponse` will always contain a timestamp, though I'm not 100% sure.
___
*[pkg/sql/sqlbase/structured.go, line 1509 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoGoZWD3MdmxiUFVnZV:-LoGoZWD3MdmxiUFVnZW:b-3cqeng) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae3b2a00366d69614cdea0c2bffac5655b1fb4d/pkg/sql/sqlbase/structured.go#L1509)):*
> ```Go
> 	// transaction will commit at the CommitTimestamp().
> 	//
> 	// TODO(vivek): Stop needing to do this by deprecating the
> ```

You're addressing this TODO in this PR. Is there anything left to do afterwards? 
___
*[pkg/sql/sqlbase/structured.go, line 3722 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoGp77yDq4dRyh9JkLt:-LoGp77yDq4dRyh9JkLu:bcs3gfn) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae3b2a00366d69614cdea0c2bffac5655b1fb4d/pkg/sql/sqlbase/structured.go#L3722)):*
> ```Go
> 			"its MVCC timestamp: has %v, expected less than %v %+v",
> 			desc.ModificationTime, ts, desc)
> 	}
> ```

I was expecting `desc.ModificationTime` to be set to `ts` if `desc.ModificationTime < ts`. It took me a moment to understand that `desc.ModificationTime.IsEmpty()` indicates that the modification time should be set from the MVCC timestamp and any non-empty values means that it was set by `MaybeIncrementVersion`. I see this is all explained in the comment for `ModificationTime`, but I just hadn't reached that part of the review yet. Wouldn't hurt to copy some of those words here as well.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40581)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang and @petermattis)

___
*[pkg/ccl/changefeedccl/table_history.go, line 286 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoGqJYg65Avlfy4sWVn:-LoGrZpE-dTJiXKB87Ga:bykj22b) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae3b2a00366d69614cdea0c2bffac5655b1fb4d/pkg/ccl/changefeedccl/table_history.go#L286)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

How bad is it going to be if you miss one of the places that table descriptors are decoded? I'm wondering if there is anything that could be done to make this less error prone.  For example, could we change the signature for `desc.GetTable()` to pass in a timestamp which would force you to update all of the callers. Seems like there are too many call sites for that, but maybe that sparks a better idea.
</blockquote></details>

It can be pretty bad. I missed the case in ConditionalGetTableDescFromTxn and it made the test fail. That specific method probably doesn't need to decode the table descriptor itself and could probably use one of the helpers from that package. 

In an earlier iteration I attempted to rename the proto `TableDescriptor` to `RawTableDescriptor` and then have a `TableDescriptor` type that wraps the RawTableDescriptor to enforce that the modification time was properly updated. That turned out to be pretty heavy and I was making a mess and preventing myself from getting to a place where I could understand if this change works. Maybe it is worth walking down that path. 

I don't know to how hard it would be to write a linter that checks if any function unmarshals into a TableDescriptor or Descriptor and then doesn't call `MaybeSetModificationTimeFromMVCCTimestamp` before returning. I guess now could be a good time to try out https://godoc.org/golang.org/x/tools/go/analysis
___
*[pkg/internal/client/db.go, line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoGmq0l9xKHMzAfH2e5:-LoGsb0096a4zhmEqKvH:b-t78j3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae3b2a00366d69614cdea0c2bffac5655b1fb4d/pkg/internal/client/db.go#L34)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

When using `ScanFormat.BATCH_RESPONSE`, the `ScanResponse.BatchResponses` field will contain encoded key/value pairs which do not have a timestamp. See `enginepb.ScanDecodeKeyValueNoTS`. I think the value in `GetResponse` will always contain a timestamp, though I'm not 100% sure.
</blockquote></details>

Yeah I noticed that. We only ever seem to use that in the KV fetcher here https://github.com/cockroachdb/cockroach/blob/1ad55c576d9786ffbd100a83e7a325389c00b8df/pkg/sql/row/kv_fetcher.go#L57 and we don't ever fetch these things as sql rows so it should be okay. I'll update the comment to explain when this is zero.
how
___
*[pkg/sql/sqlbase/structured.go, line 1509 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoGoZWD3MdmxiUFVnZV:-LoGvSGz4_D74hB8hbKS:b-ss7tb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae3b2a00366d69614cdea0c2bffac5655b1fb4d/pkg/sql/sqlbase/structured.go#L1509)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You're addressing this TODO in this PR. Is there anything left to do afterwards? 
</blockquote></details>

Nope. Done.
___
*[pkg/sql/sqlbase/structured.go, line 3722 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoGp77yDq4dRyh9JkLt:-LoGt0AW-UknxQhRTaqe:b141416) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae3b2a00366d69614cdea0c2bffac5655b1fb4d/pkg/sql/sqlbase/structured.go#L3722)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I was expecting `desc.ModificationTime` to be set to `ts` if `desc.ModificationTime < ts`. It took me a moment to understand that `desc.ModificationTime.IsEmpty()` indicates that the modification time should be set from the MVCC timestamp and any non-empty values means that it was set by `MaybeIncrementVersion`. I see this is all explained in the comment for `ModificationTime`, but I just hadn't reached that part of the review yet. Wouldn't hurt to copy some of those words here as well.
</blockquote></details>

I initially wanted it to be the case that if ModificationTime was non-zero then it would have to be exactly equal to the MVCC timestamp. Unfortunately there are places where we update these descriptors in place without touching the version or modification time (schema leases being the best example). Any time we do that now it will upgrade the descriptor to explicitly knowing about its ModificationTime. 

This way ModificationTime will either be zero and we know that it's the first value for this version and its MVCC timestamp can be used or it's non-zero and either it was written by an older version of crdb or we've updated it in place. I've add more commentary. 


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40581)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, and @petermattis)

___
*[pkg/ccl/changefeedccl/table_history.go, line 286 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoGqJYg65Avlfy4sWVn:-LoL8DvG1SGaZLb9mFzT:b-exqi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae3b2a00366d69614cdea0c2bffac5655b1fb4d/pkg/ccl/changefeedccl/table_history.go#L286)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

It can be pretty bad. I missed the case in ConditionalGetTableDescFromTxn and it made the test fail. That specific method probably doesn't need to decode the table descriptor itself and could probably use one of the helpers from that package. 

In an earlier iteration I attempted to rename the proto `TableDescriptor` to `RawTableDescriptor` and then have a `TableDescriptor` type that wraps the RawTableDescriptor to enforce that the modification time was properly updated. That turned out to be pretty heavy and I was making a mess and preventing myself from getting to a place where I could understand if this change works. Maybe it is worth walking down that path. 

I don't know to how hard it would be to write a linter that checks if any function unmarshals into a TableDescriptor or Descriptor and then doesn't call `MaybeSetModificationTimeFromMVCCTimestamp` before returning. I guess now could be a good time to try out https://godoc.org/golang.org/x/tools/go/analysis
</blockquote></details>

Well, there are only 46 callers of `GetTable()`, perhaps my idea of changing the signature isn't too onerous.
___
*[pkg/sql/sqlbase/structured.go, line 3718 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoLAK-L4OhP6LagoXRE:-LoLAK-L4OhP6LagoXRF:b-39ehd5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6aaf12b6356dda40310cf2ff4dd9f0f746fc2a7/pkg/sql/sqlbase/structured.go#L3718)):*
> ```Go
> //
> // TODO(ajwerner): devise a mechanism to ensure that this method is always
> // called when decoding table descriptors.
> ```

+1 to the TODO. 


<!-- Sent from Reviewable.io -->

Commit:
roachlint: add linter to prevent calls to sqlbase.Descriptor.GetTable()

Callers should use sqlbase.Descriptor.Table() instead.

Release justification: This is a linter for the first commit which is an
approved change.

Release note: None

ajwerner reviewed (commented):
I added a new accessor method for the TableDescriptor which takes a timestamp and validates it. There's a heavy linter too, I wanted to make it do more but alas. RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-:-LoVttZT1_hsJMaYu3k5:bbyaoei)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, and @petermattis)

___
*[pkg/ccl/backupccl/backup.go, line 307 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoVruNuEfSC8SlNff6u:-LoVruNv1JPbPn2ymxdM:b-2kdqkn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c10e6e49ef6be385014d174870fc66da44ceeed1/pkg/ccl/backupccl/backup.go#L307)):*
> ```Go
> 				"%s: unable to unmarshal SQL descriptor", row.Key)
> 		}
> 		if row.Value != nil {
> ```

This is the case I'd love a linter to catch but I suspect we'd find it with the assert when the timestamp wasn't set and the caller didn't have a reasonable timestamp. 
___
*[pkg/ccl/changefeedccl/table_history.go, line 286 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoGqJYg65Avlfy4sWVn:-LoVs5AhDwDQgGiH1GF2:b-4qu7jp) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae3b2a00366d69614cdea0c2bffac5655b1fb4d/pkg/ccl/changefeedccl/table_history.go#L286)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Well, there are only 46 callers of `GetTable()`, perhaps my idea of changing the signature isn't too onerous.
</blockquote></details>

`GetTable()` is generated. I added `Table()` that works this way and a linter to avoid calls of `GetTable()`. The linter was written in a way too heavy-weight way, I'm sure that something like the grep used in `TestTimeutil` could have worked. I had grander visions of detecting unmarshaling and requiring a call to `Table()`. 
___
*[pkg/internal/client/db.go, line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoGmq0l9xKHMzAfH2e5:-LoVt_Lj1TjfrkyBf4j3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae3b2a00366d69614cdea0c2bffac5655b1fb4d/pkg/internal/client/db.go#L34)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Yeah I noticed that. We only ever seem to use that in the KV fetcher here https://github.com/cockroachdb/cockroach/blob/1ad55c576d9786ffbd100a83e7a325389c00b8df/pkg/sql/row/kv_fetcher.go#L57 and we don't ever fetch these things as sql rows so it should be okay. I'll update the comment to explain when this is zero.
how
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner commented:
I'm extending this first commit to follow the same pattern for `CreateAsOfTime`.

thoszhang reviewed (commented):
I didn't look at the new linter closely.

> Reviewed 1 of 13 files at r1, 2 of 13 files at r2, 23 of 28 files at r3, 12 of 17 files at r5, 2 of 8 files at r7, 1 of 5 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-:-LoWet7l2BdaHznb66SI:by2qbiu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @BramGruneir, @lucy-zhang, and @petermattis)

___
*[pkg/ccl/backupccl/restore.go, line 48 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoWHgMq8KG4whP8uOVv:-LoWHgMq8KG4whP8uOVw:b-mmahgy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9fe41879b3d85b5d237cc49f50543d1b94997d0a/pkg/ccl/backupccl/restore.go#L48)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
> 	"github.com/cockroachdb/errors"
> 	opentracing "github.com/opentracing/opentracing-go"
> ```

nit: Is this intentional?
___
*[pkg/sql/conn_executor_exec.go, line 489 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoWSvGf9OLFALlq9J7Q:-LoWSvGg9Ek9yRS10j76:be5sb1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/75f52a3281ecea9d35456987023d3521f00056f6/pkg/sql/conn_executor_exec.go#L489)):*
> ```Go
> 	// with:
> 	//
> 	//  1. Leases at version V-2 for a descriptor being modified to version
> ```

I think this sentence got slightly mangled.
___
*[pkg/sql/conn_executor_exec.go, line 507 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoWQpW6APlmJ2WCyRx8:-LoWQpW6APlmJ2WCyRx9:b-tnxwsw) ([raw file](https://github.com/cockroachdb/cockroach/blob/75f52a3281ecea9d35456987023d3521f00056f6/pkg/sql/conn_executor_exec.go#L507)):*
> ```Go
> 	{
> 		// If there are existing leases at V-2 for a descriptor being modified at
> 		// version V being help, the wait loop below that waits on a cluster wide
> ```

`held`
___
*[pkg/sql/conn_executor_exec.go, line 536 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoWhAPi4IFwpDFlSDWb:-LoWhAPi4IFwpDFlSDWc:bawl9ue) ([raw file](https://github.com/cockroachdb/cockroach/blob/75f52a3281ecea9d35456987023d3521f00056f6/pkg/sql/conn_executor_exec.go#L536)):*
> ```Go
> 				filteredLeases = append(filteredLeases, l)
> 			} else if err := ex.planner.LeaseMgr().Release(l); err != nil {
> 				log.Warning(ctx, err)
> ```

What happens if we fail to release the lease here? Does it get caught in `CountLeases` and cause a retry? Do we wait somewhere else until it expires?
___
*[pkg/sql/logictest/testdata/logic_test/schema_change_retry, line 3 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoW_Fjf3EYBZKrrSyNH:-LoW_Fjf3EYBZKrrSyNI:b-j9d4da) ([raw file](https://github.com/cockroachdb/cockroach/blob/75f52a3281ecea9d35456987023d3521f00056f6/pkg/sql/logictest/testdata/logic_test/schema_change_retry#L3)):*
> ```
> # This test reproduces https://github.com/cockroachdb/cockroach/issues/23979
> 
> # Schema changes that experienced retriable errors in RELEASE
> ```

nit: This comment should be updated


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
Just saw some test failures in teamcity related to the foreign key upgrade.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-:-LoWpWbWCTVZcgAz6fkj:b3q1zo2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @BramGruneir, @lucy-zhang, and @petermattis)

___
*[pkg/sql/logictest/testdata/logic_test/schema_change_retry, line 1 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoWjHmuEBH8ge9iHwSa:-LoWjHmuEBH8ge9iHwSb:bsaysew) ([raw file](https://github.com/cockroachdb/cockroach/blob/1056a1ff962b55e046f093841fb17d7e43584960/pkg/sql/logictest/testdata/logic_test/schema_change_retry#L1)):*
> ```
> # This test reproduces https://github.com/cockroachdb/cockroach/issues/23979
> ```

I just saw the test failure under `local-mixed-19.1-19.2`. I think this test just shouldn't run with that config (which simulates a cluster running a 19.2 binary where the cluster version is an earlier version, introduced during the FK table desc upgrade), since the behavior is actually different between the cluster versions.

`# LogicTest: local fakedist fakedist-metadata fakedist-disk` is the config we use for `schema_change_in_txn` and should be fine for this too. If you think it's important to also run this test in `local-mixed-19.1-19.2`, the solution we came up with was to have a separate file that only runs in that config (e.g., `schema_change_in_txn-local-mixed-19.1-19.2`).
___
*[pkg/sql/sqlbase/structured.go, line 797 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoWo5C51BXVDyv3bfQe:-LoWo5C51BXVDyv3bfQf:b-jl8lhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/474d9d23d2b4187c69b72741d0c4ba3c653d9862/pkg/sql/sqlbase/structured.go#L797)):*
> ```Go
> 	ctx context.Context, key interface{}, msg protoutil.Message,
> ) (hlc.Timestamp, error) {
> 	return hlc.Timestamp{}, m.getProto(ctx, key, msg)
> ```

I think the `TestUpgradeDowngradeFKRepr` panic is caused by this function not returning a valid timestamp and the test itself having table descriptors that don't have the `ModificationTime` set. The more realistic test would be to somehow have this method return a valid timestamp, but an easier fix would be to set the `ModificationTime`s.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> Just saw some test failures in teamcity related to the foreign key upgrade.

fixed. 

This should be close. @andreimatei let me know if you want to take a look. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-:-LoX5Dz_2V4ewHEVDzi8:b-dyfez5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @BramGruneir, @lucy-zhang, and @petermattis)

___
*[pkg/sql/conn_executor_exec.go, line 489 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoWSvGf9OLFALlq9J7Q:-LoXCWNl7wF4beWC34EX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/75f52a3281ecea9d35456987023d3521f00056f6/pkg/sql/conn_executor_exec.go#L489)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I think this sentence got slightly mangled.
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor_exec.go, line 507 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoWQpW6APlmJ2WCyRx8:-LoXC_nP6dP6SXBUZwQZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/75f52a3281ecea9d35456987023d3521f00056f6/pkg/sql/conn_executor_exec.go#L507)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

`held`
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor_exec.go, line 536 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoWhAPi4IFwpDFlSDWb:-LoXAZi_1WTksJ4pqWDr:b-lyr6nx) ([raw file](https://github.com/cockroachdb/cockroach/blob/75f52a3281ecea9d35456987023d3521f00056f6/pkg/sql/conn_executor_exec.go#L536)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

What happens if we fail to release the lease here? Does it get caught in `CountLeases` and cause a retry? Do we wait somewhere else until it expires?
</blockquote></details>

It's not clear to me why it would fail or what it would mean. This is what the logic under `releaseLeases` did before so it can't be worse than that. My guess is yes, if it fails we'll fail and then wait for them to expire in the loop below. The error could be an ambiguous result so we may as well proceed to the check below. 
___
*[pkg/sql/logictest/testdata/logic_test/schema_change_retry, line 1 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoWjHmuEBH8ge9iHwSa:-LoX7w_6-Z1YHelRWX86:bi2e140) ([raw file](https://github.com/cockroachdb/cockroach/blob/1056a1ff962b55e046f093841fb17d7e43584960/pkg/sql/logictest/testdata/logic_test/schema_change_retry#L1)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I just saw the test failure under `local-mixed-19.1-19.2`. I think this test just shouldn't run with that config (which simulates a cluster running a 19.2 binary where the cluster version is an earlier version, introduced during the FK table desc upgrade), since the behavior is actually different between the cluster versions.

`# LogicTest: local fakedist fakedist-metadata fakedist-disk` is the config we use for `schema_change_in_txn` and should be fine for this too. If you think it's important to also run this test in `local-mixed-19.1-19.2`, the solution we came up with was to have a separate file that only runs in that config (e.g., `schema_change_in_txn-local-mixed-19.1-19.2`).
</blockquote></details>

So I modified this test because it didn't really make sense anymore. I don't have strong feeling about it. I've added the annotation you suggested.

I need to work harder to create a serializable retry. For now to make sure that the tests pass I'm removing this but I can add it back when fixed with the actual retry after. I just want to get this review into a reasonable state before I head out this evening. 
___
*[pkg/sql/logictest/testdata/logic_test/schema_change_retry, line 3 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoW_Fjf3EYBZKrrSyNH:-LoX8929B5my_ljtT5GZ:blbizx6) ([raw file](https://github.com/cockroachdb/cockroach/blob/75f52a3281ecea9d35456987023d3521f00056f6/pkg/sql/logictest/testdata/logic_test/schema_change_retry#L3)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

nit: This comment should be updated
</blockquote></details>

Sure, but what should it say.
___
*[pkg/sql/sqlbase/structured.go, line 3718 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoLAK-L4OhP6LagoXRE:-LoXBq_15paJFbbTg7Wq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6aaf12b6356dda40310cf2ff4dd9f0f746fc2a7/pkg/sql/sqlbase/structured.go#L3718)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

+1 to the TODO. 
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.go, line 797 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoWo5C51BXVDyv3bfQe:-LoXBCKU8Mq9jX5XnsR2:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/474d9d23d2b4187c69b72741d0c4ba3c653d9862/pkg/sql/sqlbase/structured.go#L797)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I think the `TestUpgradeDowngradeFKRepr` panic is caused by this function not returning a valid timestamp and the test itself having table descriptors that don't have the `ModificationTime` set. The more realistic test would be to somehow have this method return a valid timestamp, but an easier fix would be to set the `ModificationTime`s.
</blockquote></details>

Done. 


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40581)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @BramGruneir, @lucy-zhang, and @petermattis)

___
*[pkg/internal/client/db.go, line 338 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoXTW133Ra9jS46yPFW:-LoXTW133Ra9jS46yPFX:bb92sec) ([raw file](https://github.com/cockroachdb/cockroach/blob/e843c54f9ca82480930c780eb295f84192e7c7b4/pkg/internal/client/db.go#L338)):*
> ```Go
> func (db *DB) GetProtoTs(
> 	ctx context.Context, key interface{}, msg protoutil.Message,
> ) (ts hlc.Timestamp, err error) {
> ```

nit: get rid of the names on the return values. They don't add anything. And they invite a nit on the `return ts, err` line about how it should be `return hlc.Timestamp{}, err`
___
*[pkg/sql/conn_executor_exec.go, line 477 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoXVQOe4e8mXPs0aTgd:-LoXVQOe4e8mXPs0aTge:b4jty8a) ([raw file](https://github.com/cockroachdb/cockroach/blob/e843c54f9ca82480930c780eb295f84192e7c7b4/pkg/sql/conn_executor_exec.go#L477)):*
> ```Go
> // proceed.
> func (ex *connExecutor) checkTableTwoVersionInvariant(ctx context.Context) error {
> 	tables := ex.extraTxnState.tables.getTablesWithNewVersion()
> ```

nit: move this `tables` below the big comments
___
*[pkg/sql/conn_executor_exec.go, line 490 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoXUWOA9Cjk9uxXwZ_m:-LoXUWOA9Cjk9uxXwZ_n:bim8ice) ([raw file](https://github.com/cockroachdb/cockroach/blob/e843c54f9ca82480930c780eb295f84192e7c7b4/pkg/sql/conn_executor_exec.go#L490)):*
> ```Go
> 	//
> 	//  1. Leases at version V-2 for a descriptor being modified to version V.
> 	//  These leases must be dropped immediately because we're going to assert
> ```

nit: we're not going to "assert" anything, we're going to "check" something. Saying assert seems to contradict the next sentence.
___
*[pkg/sql/conn_executor_exec.go, line 505 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoXUwSf1ftu2SR_aYFD:-LoXUwSf1ftu2SR_aYFE:bb5wx4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/e843c54f9ca82480930c780eb295f84192e7c7b4/pkg/sql/conn_executor_exec.go#L505)):*
> ```Go
> 	// cases explicitly. We must release the 1. leases above before we attempt
> 	// to check the invariant, otherwise it's doomed to fail.
> 	// the other leases we need to wait until we commit to release.
> ```

s/the/The
___
*[pkg/sql/conn_executor_exec.go, line 512 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoXVdYZAAOjdnwfOm1d:-LoXVdY_AWjemafmdAku:bdjm236) ([raw file](https://github.com/cockroachdb/cockroach/blob/e843c54f9ca82480930c780eb295f84192e7c7b4/pkg/sql/conn_executor_exec.go#L512)):*
> ```Go
> 		// version V being held, the wait loop below that waits on a cluster wide
> 		// release of old version leases will hang until these leases expire so
> 		// we need to drop these leases right now. It is not safe to release leases
> ```

can you expand on this "it is not safe to release..."? I assume it has to do with the fact that the current txn might have used them. But how come we're sure it hasn't used the v-2 ones? I believe that's probably true, but I'd like to see the argument.

And actually... Are you sure we can't release the V-1 ones too? What's the hazard? The hazard, I guess, is that this transaction would then be free to commit at a timestamp above the lease (which generally is not allowed because that might be outside of V-1's validity). *But*, if the transaction commits, it is writing V (or perhaps another transaction will write V, but definitely at a timestamp above the commit ts). So then, I think it is in fact safe to commit the current transaction, because we know that V-1's validity is being extended.

I don't know what I'm talking about, although I'm probably "the expert" on this stuff. But it's been a while and you seem to have done your homework, so perhaps you can school me.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Thanks for the reviews! RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-:-LoYkxgD29S0CuLmiZ5x:b-21u8wv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @BramGruneir, @lucy-zhang, and @petermattis)

___
*[pkg/internal/client/db.go, line 338 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoXTW133Ra9jS46yPFW:-LoYUwUx357a0Y9W6OYK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e843c54f9ca82480930c780eb295f84192e7c7b4/pkg/internal/client/db.go#L338)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: get rid of the names on the return values. They don't add anything. And they invite a nit on the `return ts, err` line about how it should be `return hlc.Timestamp{}, err`
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor_exec.go, line 477 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoXVQOe4e8mXPs0aTgd:-LoYYpn26VBtS3ehwbUz:b4spxpi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e843c54f9ca82480930c780eb295f84192e7c7b4/pkg/sql/conn_executor_exec.go#L477)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: move this `tables` below the big comments
</blockquote></details>

I feel like the short circuit coming first is reasonable. This method gets called on every commit and this is the the common case.

I reworked the comments moved the lease releasing to a method too. 
___
*[pkg/sql/conn_executor_exec.go, line 490 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoXUWOA9Cjk9uxXwZ_m:-LoYV2LU-li4FOiJxM8J:b-qtn6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/e843c54f9ca82480930c780eb295f84192e7c7b4/pkg/sql/conn_executor_exec.go#L490)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: we're not going to "assert" anything, we're going to "check" something. Saying assert seems to contradict the next sentence.
</blockquote></details>

"assert" has been scrubbed from the comment. 
___
*[pkg/sql/conn_executor_exec.go, line 505 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoXUwSf1ftu2SR_aYFD:-LoYbpNT2Z5Jq1RhgkGR:b-32gzc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e843c54f9ca82480930c780eb295f84192e7c7b4/pkg/sql/conn_executor_exec.go#L505)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/the/The
</blockquote></details>

reworked. 
___
*[pkg/sql/conn_executor_exec.go, line 512 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoXVdYZAAOjdnwfOm1d:-LoYV4HN5Ezbw-ZRaQmN:b-1erkpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e843c54f9ca82480930c780eb295f84192e7c7b4/pkg/sql/conn_executor_exec.go#L512)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can you expand on this "it is not safe to release..."? I assume it has to do with the fact that the current txn might have used them. But how come we're sure it hasn't used the v-2 ones? I believe that's probably true, but I'd like to see the argument.

And actually... Are you sure we can't release the V-1 ones too? What's the hazard? The hazard, I guess, is that this transaction would then be free to commit at a timestamp above the lease (which generally is not allowed because that might be outside of V-1's validity). *But*, if the transaction commits, it is writing V (or perhaps another transaction will write V, but definitely at a timestamp above the commit ts). So then, I think it is in fact safe to commit the current transaction, because we know that V-1's validity is being extended.

I don't know what I'm talking about, although I'm probably "the expert" on this stuff. But it's been a while and you seem to have done your homework, so perhaps you can school me.
</blockquote></details>

You're right that this is too complicated. The intents on the table descriptors ensure that V-1's validity will continue until this transaction commits. The thing we need to be careful of is to not release are the leases for tables we're not modifying, that was previously safe because the timestamp was fixed but no more.

I've updated the code to drop all leases for tables which are being modified.
___
*[pkg/sql/logictest/testdata/logic_test/schema_change_retry, line 1 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoWjHmuEBH8ge9iHwSa:-LoYdJNM1MiN0Buuk_VW:buu63lo) ([raw file](https://github.com/cockroachdb/cockroach/blob/1056a1ff962b55e046f093841fb17d7e43584960/pkg/sql/logictest/testdata/logic_test/schema_change_retry#L1)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

So I modified this test because it didn't really make sense anymore. I don't have strong feeling about it. I've added the annotation you suggested.

I need to work harder to create a serializable retry. For now to make sure that the tests pass I'm removing this but I can add it back when fixed with the actual retry after. I just want to get this review into a reasonable state before I head out this evening. 
</blockquote></details>

Okay I made it make sense again with the use of 
```
SET CLUSTER SETTING kv.transaction.max_refresh_spans_bytes = 0;   
```

Done.
___
*[pkg/sql/sqlbase/structured.go, line 3722 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoGp77yDq4dRyh9JkLt:-LoYdVJX41XleqxIeFJC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae3b2a00366d69614cdea0c2bffac5655b1fb4d/pkg/sql/sqlbase/structured.go#L3722)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I initially wanted it to be the case that if ModificationTime was non-zero then it would have to be exactly equal to the MVCC timestamp. Unfortunately there are places where we update these descriptors in place without touching the version or modification time (schema leases being the best example). Any time we do that now it will upgrade the descriptor to explicitly knowing about its ModificationTime. 

This way ModificationTime will either be zero and we know that it's the first value for this version and its MVCC timestamp can be used or it's non-zero and either it was written by an older version of crdb or we've updated it in place. I've add more commentary. 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

thoszhang reviewed (approved):
> Reviewed 1 of 11 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40581)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @BramGruneir, @lucy-zhang, and @petermattis)

___
*[pkg/sql/table.go, line 389 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LobVlkS03U9xXBp1x42:-LobVlkS03U9xXBp1x43:b-zfb75t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e5ed74573bdf5165e772aba48e2a503e6c9132a/pkg/sql/table.go#L389)):*
> ```Go
> }
> 
> // releaseModifiedTableLeases releases the leases for the tables with ids in
> ```

`releaseTableLeases` (or maybe rename the method)
___
*[pkg/sql/table.go, line 393 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LocJpMy3AbKnhDFarFi:-LocJpMy3AbKnhDFarFj:b-5si3i3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e5ed74573bdf5165e772aba48e2a503e6c9132a/pkg/sql/table.go#L393)):*
> ```Go
> func (tc *TableCollection) releaseTableLeases(ctx context.Context, tables []IDVersion) {
> 	// Sort the tables and leases to make it easy to find the leases to release.
> 	leasedTables := tc.leasedTables
> ```

Just so I understand what's going on: Assuming a table is being upgraded to version V in this transaction (implying, I think, that its version in `tables` is V-2), can `leasedTables` contain table descs at versions V-1 *and* V-2?
___
*[pkg/sql/logictest/testdata/logic_test/schema_change_retry, line 1 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoWjHmuEBH8ge9iHwSa:-LocKi1q7fG3Gb-vvPGd:bf0240n) ([raw file](https://github.com/cockroachdb/cockroach/blob/1056a1ff962b55e046f093841fb17d7e43584960/pkg/sql/logictest/testdata/logic_test/schema_change_retry#L1)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Okay I made it make sense again with the use of 
```
SET CLUSTER SETTING kv.transaction.max_refresh_spans_bytes = 0;   
```

Done.
</blockquote></details>

My earlier comment was assuming that the updated test was intended to not have an error at all, in which case the behavior would be different in `local` and `local-mixed-19.1-19.2`, but you're right that that doesn't make sense as a test.


If this test passes in `local-mixed-19.1-19.2` as-is, I don't see any reason to not run it, so these config lines should be taken out.
___
*[pkg/sql/logictest/testdata/logic_test/schema_change_retry, line 3 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoW_Fjf3EYBZKrrSyNH:-LocKU-E2R7cbL5gT2og:b-v39elz) ([raw file](https://github.com/cockroachdb/cockroach/blob/75f52a3281ecea9d35456987023d3521f00056f6/pkg/sql/logictest/testdata/logic_test/schema_change_retry#L3)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Sure, but what should it say.
</blockquote></details>

Never mind, see below.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

Really happy about this. It's been a dream for years.

> Reviewed 1 of 13 files at r2, 1 of 28 files at r3, 1 of 17 files at r5, 2 of 10 files at r13.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-:-Lofgb7R-QdgLDHJCV5w:bgrnm1r)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @BramGruneir, @lucy-zhang, and @petermattis)

___
*[pkg/settings/cluster/cockroach_versions.go, line 537 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LofdMYV8LJAxjH73066:-LofdMYV8LJAxjH73067:bb7pa6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/settings/cluster/cockroach_versions.go#L537)):*
> ```Go
> 	},
> 	{
> 		// VersionTableDescModificationTimeFromMVCC is TODO
> ```

todo
___
*[pkg/settings/cluster/cockroach_versions.go, line 539 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LofdQot1tIRH0iDCzPT:-LofdQot1tIRH0iDCzPU:b-ch1l77) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/settings/cluster/cockroach_versions.go#L539)):*
> ```Go
> 		// VersionTableDescModificationTimeFromMVCC is TODO
> 		//
> 		// It represents an upgrade to the table descriptor format in which
> ```

please spell it out more. Say what fields from what proto are no longer populated in storage
___
*[pkg/sql/conn_executor_exec.go, line 477 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoXVQOe4e8mXPs0aTgd:-LofgQlPChnAyGzD51uB:b-hxex1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/e843c54f9ca82480930c780eb295f84192e7c7b4/pkg/sql/conn_executor_exec.go#L477)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I feel like the short circuit coming first is reasonable. This method gets called on every commit and this is the the common case.

I reworked the comments moved the lease releasing to a method too. 
</blockquote></details>

oh I missed the short-circuit, nvm
___
*[pkg/sql/plan.go, line 71 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-Lofel-11_3dsETsvoPt:-Lofel-11_3dsETsvoPu:b-wjtx28) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/sql/plan.go#L71)):*
> ```Go
> //
> // TODO(ajwerner): remove in 20.1.
> func (r *runParams) creationTimeForNewTableDescriptor() hlc.Timestamp {
> ```

This strikes me as an unfortunate method to have. Have you considered modifying `InitTableDescriptor` (as a choke point) to look at the cluster version (and modifying its signature to take a `txn` instead of a timestamp)?
Don't want to create work for you though, so feel free to ignore if you're not interested.
___
*[pkg/sql/plan.go, line 74 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LofdqSd80FVo85B9vnu:-LofdqSd80FVo85B9vnv:b-szafvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/sql/plan.go#L74)):*
> ```Go
> 	// Before 19.2 we needed to observe the transaction CommitTimestamp to ensure
> 	// that CreateAsOfTime and ModificationTime reflected the timestamp at which the
> 	// creating transaction	committed. Starting in 19.2 we use a zero-valued
> ```

nit: double space
___
*[pkg/sql/plan.go, line 76 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-Lofdwc915Vl80diHHV0:-Lofdwc915Vl80diHHV1:bm0wn7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/sql/plan.go#L76)):*
> ```Go
> 	// creating transaction	committed. Starting in 19.2 we use a zero-valued
> 	// CreateAsOfTime and ModificationTime when creating a table descriptor and then
> 	// upon reading use the MVCC timestamp to	populate the values.
> ```

double space
___
*[pkg/sql/table.go, line 398 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-Lofc1F-BIYETZgio-wb:-Lofc1F-BIYETZgio-wc:b-yo00m) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/sql/table.go#L398)):*
> ```Go
> 			return tables[i].version < tables[j].version
> 		}
> 		return tables[i].id < tables[j].id
> ```

nit: invert the order in which we look at fields
___
*[pkg/sql/table.go, line 413 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-Lofctgk96HzlqgvNy71:-Lofctgk96HzlqgvNy72:b-hffai8) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/sql/table.go#L413)):*
> ```Go
> 			tablesToConsider = tablesToConsider[1:]
> 		}
> 		return len(tablesToConsider) > 0 && tablesToConsider[0].id == id
> ```

I'm confused; I don't see where we check the version from the input `tables`.
___
*[pkg/sql/sqlbase/structured.proto, line 531 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LofbJ2KC_taf090Iqf7:-LofbJ2KC_taf090Iqf8:b68m54z) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/sql/sqlbase/structured.proto#L531)):*
> ```protobuf
>   // should live inside of a Descriptor. The Descriptor.Table() method takes an
>   // hlc timestamp to ensure that this field is set properly when extracted from
>   // a Descriptor.
> ```

So should there be a TODO here for 20.1 to get rid of this field? Is there any hope that the field can live on another structures (perhaps the `ImmutableTableDescriptor`)?
___
*[pkg/sql/sqlbase/structured.proto, line 822 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LofbXMs8e4Tx2UwQ5Qc:-LofbXMs8e4Tx2UwQ5Qd:b-kjrh5a) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/sql/sqlbase/structured.proto#L822)):*
> ```protobuf
> 
>   optional string create_query = 34 [(gogoproto.nullable) = false];
>      
> ```

spaces
___
*[pkg/sql/sqlbase/structured.proto, line 825 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-Lofb_fKE7t0QzMCzDo5:-Lofb_fKE7t0QzMCzDo6:b-lre6zj) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/sql/sqlbase/structured.proto#L825)):*
> ```protobuf
>   // Starting in 19.2 CreateAsOfTime is initialized to zero for the first
>   // version of a table and is populated from the MVCC timestamp of the read
>   // like ModificationTime. See Descriptor.Table().
> ```

ditto TODO


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40581)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @BramGruneir, @lucy-zhang, and @petermattis)

___
*[pkg/settings/cluster/cockroach_versions.go, line 537 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LofdMYV8LJAxjH73066:-LogWPLk47dE8_eicdT0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/settings/cluster/cockroach_versions.go#L537)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

todo
</blockquote></details>

Done.
___
*[pkg/settings/cluster/cockroach_versions.go, line 539 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LofdQot1tIRH0iDCzPT:-LogWPwX0XZjErwOjmJi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/settings/cluster/cockroach_versions.go#L539)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please spell it out more. Say what fields from what proto are no longer populated in storage
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor_exec.go, line 536 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoWhAPi4IFwpDFlSDWb:-LogWUjD3XoLMStDbiyf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/75f52a3281ecea9d35456987023d3521f00056f6/pkg/sql/conn_executor_exec.go#L536)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

It's not clear to me why it would fail or what it would mean. This is what the logic under `releaseLeases` did before so it can't be worse than that. My guess is yes, if it fails we'll fail and then wait for them to expire in the loop below. The error could be an ambiguous result so we may as well proceed to the check below. 
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor_exec.go, line 505 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoXUwSf1ftu2SR_aYFD:-LogWTETAfIQuCq2agyq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e843c54f9ca82480930c780eb295f84192e7c7b4/pkg/sql/conn_executor_exec.go#L505)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

reworked. 
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor_exec.go, line 512 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoXVdYZAAOjdnwfOm1d:-LogWTmH4pJJ0uEwMA8c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e843c54f9ca82480930c780eb295f84192e7c7b4/pkg/sql/conn_executor_exec.go#L512)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

You're right, that this is too complicated. The intents on the table descriptors ensure that V-1's validity will continue until this transaction commits. The thing we need to be careful to not release are the leases for tables we're not modifying, that was previously safe because the timestamp was fixed but no more.

I've updated the code to drop all leases for tables which are being modified.
</blockquote></details>

Done.
___
*[pkg/sql/plan.go, line 71 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-Lofel-11_3dsETsvoPt:-LogWmBoAnTiqeE6F_sP:b-a4sqd8) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/sql/plan.go#L71)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

This strikes me as an unfortunate method to have. Have you considered modifying `InitTableDescriptor` (as a choke point) to look at the cluster version (and modifying its signature to take a `txn` instead of a timestamp)?
Don't want to create work for you though, so feel free to ignore if you're not interested.
</blockquote></details>

And also a cluster version? We can remove this soon enough. I think I'm going to pass. 
___
*[pkg/sql/plan.go, line 76 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-Lofdwc915Vl80diHHV0:-LogWxtb4rcYIe3pnski:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/sql/plan.go#L76)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

double space
</blockquote></details>

Done.
___
*[pkg/sql/table.go, line 398 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-Lofc1F-BIYETZgio-wb:-LogXh22ADhJm2RhX684:b-j5u8tg) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/sql/table.go#L398)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: invert the order in which we look at fields
</blockquote></details>

I've removed version. 
___
*[pkg/sql/table.go, line 413 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-Lofctgk96HzlqgvNy71:-LogXedK228JKq32kae-:btf8cgj) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/sql/table.go#L413)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'm confused; I don't see where we check the version from the input `tables`.
</blockquote></details>

Nothing uses version anymore. I've stopped sorting by it. 
___
*[pkg/sql/table.go, line 389 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LobVlkS03U9xXBp1x42:-LogZ4oo6-MnO5D22RoU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e5ed74573bdf5165e772aba48e2a503e6c9132a/pkg/sql/table.go#L389)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

`releaseTableLeases` (or maybe rename the method)
</blockquote></details>

Done.
___
*[pkg/sql/table.go, line 393 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LocJpMy3AbKnhDFarFi:-LogX3Go1CtuTMpt3JsF:b-3q3w9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e5ed74573bdf5165e772aba48e2a503e6c9132a/pkg/sql/table.go#L393)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Just so I understand what's going on: Assuming a table is being upgraded to version V in this transaction (implying, I think, that its version in `tables` is V-2), can `leasedTables` contain table descs at versions V-1 *and* V-2?
</blockquote></details>

I don't know enough to know. The previously existing comment made me think that we could be holding V-2 leases given how problematic they would be. As for the V-1 leases, isn't that the current version? 
___
*[pkg/sql/logictest/testdata/logic_test/schema_change_retry, line 1 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LoWjHmuEBH8ge9iHwSa:-LogZUmuAfKyjBkgNW6I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1056a1ff962b55e046f093841fb17d7e43584960/pkg/sql/logictest/testdata/logic_test/schema_change_retry#L1)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

My earlier comment was assuming that the updated test was intended to not have an error at all, in which case the behavior would be different in `local` and `local-mixed-19.1-19.2`, but you're right that that doesn't make sense as a test.


If this test passes in `local-mixed-19.1-19.2` as-is, I don't see any reason to not run it, so these config lines should be taken out.
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.proto, line 531 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LofbJ2KC_taf090Iqf7:-LogYHyq-rscOVkB-xpq:bqc2n8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/sql/sqlbase/structured.proto#L531)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

So should there be a TODO here for 20.1 to get rid of this field? Is there any hope that the field can live on another structures (perhaps the `ImmutableTableDescriptor`)?
</blockquote></details>

No this field needs to continue to live. We update table descriptors without changing their version. When we re-write them we populate them with a non-zero value for this field. 
___
*[pkg/sql/sqlbase/structured.proto, line 822 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-LofbXMs8e4Tx2UwQ5Qc:-LogZeV11LRH_08mrmCr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/sql/sqlbase/structured.proto#L822)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

spaces
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.proto, line 825 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/40581#-Lofb_fKE7t0QzMCzDo5:-LogYRwzA27Bj3BBvNI8:b-ux5zij) ([raw file](https://github.com/cockroachdb/cockroach/blob/33dae63a61f667718fc488cd5df60333bfd9ee98/pkg/sql/sqlbase/structured.proto#L825)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

ditto TODO
</blockquote></details>

Nope, this field is only zero when the table is first created. For all future writes this field has a value. 


<!-- Sent from Reviewable.io -->

ajwerner commented:
TFTR!

bors r+

ajwerner commented:
Needed a release justification in the description.

bors r+
